log/context.go

```diff
@@ -6,12 +6,31 @@ import (
 
 	"github.com/getsentry/sentry-go"
 	"go.uber.org/zap"
+
+	"github.com/reddit/baseplate.go/detach"
 )
 
 type contextKeyType struct{}
 
 var contextKey contextKeyType
 
+func init() {
+	detach.Register(detach.Hooks{
+		Inline: func(dst, src context.Context) context.Context {
+			if logger, ok := src.Value(contextKey).(*zap.SugaredLogger); ok && logger != nil {
+				dst = context.WithValue(dst, contextKey, logger)
+			}
+			return dst
+		},
+		Async: func(dst, src context.Context, next func(ctx context.Context)) {
+			if logger, ok := src.Value(contextKey).(*zap.SugaredLogger); ok && logger != nil {
+				dst = context.WithValue(dst, contextKey, logger)
+			}
+			next(dst)
+		},
```

Comment:
🔕 we did the same copy in both `Inline` and `Async` here, and that would be a common case for a lot of packages (they need to copy some shared kv pairs in both `Inline` and `Async` hooks). I think we should help avoid this kind of common code duplication from the API level, for example provide a `Base` hook that will be called by both `Inline` and `Async` if it's non-nil?

Reply:
IMO, that's messier than the current state, while there's some duplication, it's minor and only happens in a few libraries. The issue I see with adding something like `Base` is it makes it less straightforward to see what's happening just by reading the `Register` call, you would have to know that if `Base` is defined, it get's called before each of these functions.

Reply:
then at least we should demonstrate a good example here to avoid code duplication :) something like:
```go
func init() {
  copyContext := func(dst, src context.Context) context.Context {
    ...
  }
  detach.Register(detach.Hooks{
    Inline: func(dst, src context.Context) context.Context {
      return copyContext(dst, src)
    },
    Async: func(dst, src context.Context, next func(ctx context.Context)) {
      next(copyContext(dst, src))
    },
  })
}
```

Reply:
random idea: what if we also add the hooks itself to `Async`'s arg list? this way it could be this instead:
```go
detach.Register(detach.Hooks{
  Inline: func(dst, src context.Context) context.Context {
    ...
  },
  Async: func(dst, src context.Context, hook detach.Hooks, next func(ctx context.Context)) {
    next(hook.Inline(dst, src))
  },
})
```
we can also provide that `Async` implementation from `detach` library directly as that works for a good amount of packages, so they only need to implement `Inline` and just use the pre-provided `Async`.